db/db_compaction_test.cc

```diff
@@ -6464,82 +6458,366 @@ TEST_P(DBCompactionTestWithParam,
+          "PickCompactionToReduceSizeAmpReturnNonnullptr", [&](void* /*arg*/) {
+            compaction_path_sync_point_called_.store(true);
+          });
+    } else if (compaction_path_to_test == "PickCompactionToReduceSortedRuns") {
+      assert(options_.compaction_style ==
+             CompactionStyle::kCompactionStyleUniversal);
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+          "PickCompactionToReduceSortedRunsReturnNonnullptr",
+          [&](void* /*arg*/) {
+            compaction_path_sync_point_called_.store(true);
+          });
+    } else if (compaction_path_to_test == "FindIntraL0Compaction") {
+      assert(options_.compaction_style ==
+             CompactionStyle::kCompactionStyleFIFO);
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+          "FindIntraL0Compaction", [&](void* /*arg*/) {
+            compaction_path_sync_point_called_.store(true);
+          });
+    } else if (compaction_path_to_test == "PickCompactionToWarm" ||
+               compaction_path_to_test == "CompactRange") {
+      assert(options_.compaction_style ==
+             CompactionStyle::kCompactionStyleFIFO);
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+          "PickCompactionToWarm::BeforeGetCurrentTime",
+          [this](void* current_time_arg) -> void {
+            // The unit test goes so quickly that there is almost no time
+            // elapsed after we ingest a file and before we check whether
+            // ingested files can compact to warm. Therefore we need this trick
+            // to simulate elapsed time in reality.
+            int64_t* current_time = (int64_t*)current_time_arg;
+            *current_time = *current_time +
+                            options_.compaction_options_fifo.age_for_warm + 1;
+          });
+      ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->SetCallBack(
+          "PostPickCompactionToWarm", [&](void* compaction_arg) {
+            Compaction* compaction = (Compaction*)compaction_arg;
+            if (compaction != nullptr) {
+              compaction_path_sync_point_called_.store(true);
+            }
+          });
+    }
+
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->EnableProcessing();
+  }
+
+  bool SyncPointsCalled() { return compaction_path_sync_point_called_.load(); }
+
+  void DisableSyncPoints() {
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->ClearAllCallBacks();
+    ROCKSDB_NAMESPACE::SyncPoint::GetInstance()->DisableProcessing();
+  }
+
+ protected:
+  Options options_;
+
+ private:
+  const Snapshot* snapshot_ = nullptr;
+  std::atomic<bool> compaction_path_sync_point_called_;
+};
+
+TEST_F(DBCompactionTestL0FilesReorderingCorruption,
+       FlushAfterIntraL0UniversalCompactionWithIngestedFile) {
+  for (const std::string compaction_path_to_test :
+       {"PickPeriodicCompaction", "PickCompactionToReduceSizeAmp",
+        "PickCompactionToReduceSortedRuns"}) {
```

Comment:
I tried to cover `PickDeleteTriggeredCompaction()` but only come up with low-level test case manipulating vstorage directly like https://github.com/facebook/rocksdb/blob/main/db/compaction/compaction_picker_test.cc#L2953-L2977. Is there another way of doing this? If not, I can still go with the low-level one - better than nothing.

Reply:
Actually - found another more useful test case TEST_P(DBCompactionTestWithParam, CompactionDeletionTrigger)

Reply:
Fixed